tools/pip-compile-wrapper.py

```diff
@@ -0,0 +1,300 @@
+        source_path = source_path.resolve()
+        # Dummy access to create the key
+        _ = SOURCE_DEPENDENCIES[source_name]
+        target_dir: Path = source_path.parent
+        with source_path.open("rt") as target_file:
+            for line in target_file:
+                line = line.strip()
+                if line.startswith("-r"):
+                    required = (
+                        target_dir.joinpath(line.lstrip("-r").strip())
+                        .resolve()
+                        .relative_to(REQUIREMENTS_DIR)
+                        .stem
+                    )
+                    SOURCE_DEPENDENCIES[source_name].add(required)
+
+
+_resolve_source_dependencies()
+
+
+def _run_pip_compile(
+    source_name: str,
+    upgrade_all: bool = False,
+    upgrade_packages: Optional[Set[str]] = None,
+    verbose: bool = False,
+    dry_run: bool = False,
+) -> None:
+    assert_msg = "Only one of `upgrade_all` or `upgrade_packages` may be given."
+    assert not (upgrade_all and upgrade_packages), assert_msg
+
+    pip_compile_exe = which("pip-compile")
+    if not pip_compile_exe:
+        raise RuntimeError("pip-compile missing. This shouldn't happen.")
+
+    if not upgrade_packages:
+        upgrade_packages = set()
+
+    source_path = SOURCE_PATHS[source_name]
+    target_path = TARGET_PATHS[source_name]
+
+    upgrade_packages_cmd: List[str] = []
+    if upgrade_packages:
+        packages_in_target = {
+            package_name
+            for package_name, _ in _get_requirement_packages(source_name, TargetType.ALL)
+        }
+
+        upgrade_packages_cmd = list(
+            chain.from_iterable(
+                zip(repeat("--upgrade-package"), upgrade_packages.intersection(packages_in_target))
+            )
+        )
+    upgrade_all_cmd: List[str] = []
+    if upgrade_all:
+        upgrade_all_cmd = ["--upgrade"]
+
+    dry_run_cmd = ["--dry-run"] if dry_run else []
+
+    command = [
+        pip_compile_exe,
+        "--verbose" if verbose else "--quiet",
+        *dry_run_cmd,
+        "--no-index",
+        *upgrade_packages_cmd,
+        *upgrade_all_cmd,
+        "--output-file",
+        str(target_path),
+        str(source_path),
+    ]
+    print(f"Compiling {source_path.name}...", end="")
+    sys.stdout.flush()
+    if verbose:
+        print(f"\nRunning command: {' '.join(shlex.quote(c) for c in command)}")
+    env = os.environ.copy()
+    env["CUSTOM_COMPILE_COMMAND"] = "'./deps compile' (for details see README)"
+    process = subprocess.run(
+        command, capture_output=(not verbose), cwd=str(source_path.parent), env=env
+    )
+    if process.returncode == 0:
+        print("\b\b Success.")
+        return
+    print("\b\b Error!")
+    if not verbose:
+        print(process.stdout.decode())
+        print(process.stderr.decode())
+    process.check_returncode()
+
+
+def _resolve_deps(source_names: Iterable[str]) -> List[Set[str]]:
```

Comment:
Please add a docstring, e.g.:

> Partial order `source_names` based on their dependencies, raise an Exception if not possible.
>
> The resulting list has the following property: Each entry does not depend on a *later* entry.

Reply:
I think this should return `List[str]` and not `List[Set[str]]`, there is no relation to the items in a set, and this will simplify the callers which are doing two loops.

Reply:
Yes, true. 
Initially I was thinking about running the independent items in the sub groups in parallel (since initial runs of pip-compile can take quite a while). But once its cache is primed they are fast enough to not make it worth the effort.